File: StellaServerAPI/Protocol/StringProtocol.cs
﻿using System;
using System.Text;

namespace StellaServerAPI.Protocol
{
    public class StringProtocol
    {
        /// <summary>
        /// The number of bytes needed for the header of the package.
        /// Represents:
        /// - Number of packages if it's the first package.
        /// - Package index if it's an subsequent package.
        /// </summary>
        const int PACKAGE_HEADER_BYTES = sizeof(int);

        // Convert the string to an array of byte arrays
        public static byte[][] Serialize(string message, int maxPackageSize)
        {
            byte[] data = Encoding.ASCII.GetBytes(message);

            int bytesAvailablePerPackage = maxPackageSize - PACKAGE_HEADER_BYTES;
            int packagesNeeded = (data.Length + bytesAvailablePerPackage - 1) / bytesAvailablePerPackage;
            
            byte[][] returnData = new byte[packagesNeeded][];
            // Create first package
            returnData[0] = CreatePackage(data, 0, bytesAvailablePerPackage, packagesNeeded);
            for (int i = 1; i < packagesNeeded; i++)
            {
                returnData[i] = CreatePackage(data, i * bytesAvailablePerPackage, bytesAvailablePerPackage, i);
            }
            return returnData;
        }

        /// <summary>
        ///  Create a subsequent package with a normal package header
        /// </summary>
        private static byte[] CreatePackage(byte[] data, int startIndex, int maxBodySize, int headerCounter)
        {
            int packageSize = Math.Min(maxBodySize + PACKAGE_HEADER_BYTES, data.Length - startIndex + PACKAGE_HEADER_BYTES);
            byte[] buffer = new byte[packageSize];
            BitConverter.GetBytes(headerCounter).CopyTo(buffer, 0);
            Array.Copy(data,startIndex,buffer,4,buffer.Length - 4);
            return buffer;
        }


        private StringBuilder _stringBuilder;
        private int _numberOfPackages;
        private int _packagesReceived;

        public StringProtocol()
        {
            _stringBuilder = new StringBuilder();
        }

        /// <summary>
        /// Returns true when the string has been fully deserialized
        /// </summary>
        /// <param name="package"></param>
        /// <returns></returns>
        public bool TryDeserialize(byte[] package, out string message)
        {
            message = null;
            if (_packagesReceived == 0)
            {
                // First package.
                _numberOfPackages = BitConverter.ToInt32(package, 0);
            }

            _stringBuilder.Append(Encoding.ASCII.GetString(package, 4, package.Length - 4));
            _packagesReceived++;

            if (_packagesReceived == _numberOfPackages)
            {
                message = _stringBuilder.ToString();
                return true;
            }

            return false;
        }
    }
}

File: StellaWebInterface/backend/service/stringProtocol.js
class StringProtocol {
  constructor() {
    this.PACKAGE_HEADER_BYTES = 4;
    this._stringBuilder = '';
    this._numberOfPackages = 0;
    this._packagesReceived = 0;
    this.message = null;
  }

  // Convert the string to an array of byte arrays
  serialize(message, maxPackageSize) {
    const data = new Buffer.from(message, 'ascii');

    const bytesAvailablePerPackage = maxPackageSize - this.PACKAGE_HEADER_BYTES;
    const packagesNeeded = Math.floor((data.length + bytesAvailablePerPackage - 1) / bytesAvailablePerPackage);

    const returnData = [];
    // Create first package
    returnData.push(this.createPackage(data, 0, bytesAvailablePerPackage, packagesNeeded));
    for (let i = 1; i < packagesNeeded; i++) {
      returnData.push(this.createPackage(data, i * bytesAvailablePerPackage, bytesAvailablePerPackage, i));
    }
    return returnData;
  }

  /// <summary>
  ///  Create a subsequent package with a normal package header
  /// </summary>
  createPackage(data, startIndex, maxBodySize, headerCounter) {
    const packageSize = Math.min(maxBodySize + this.PACKAGE_HEADER_BYTES, data.length - startIndex + this.PACKAGE_HEADER_BYTES);
    const buffer = new Buffer.alloc(packageSize);

    buffer.writeInt32LE(headerCounter, 0);
    data.copy(buffer, 4, startIndex, buffer.length - 4);
    return buffer;
  }

  /// <summary>
  /// Returns true when the string has been fully deserialized
  /// </summary>
  /// <param name="package"></param>
  /// <returns></returns>
  deserialize(data) {
    if (this._packagesReceived === 0) {
      this._numberOfPackages = data.readUInt32LE();
    }

    this._stringBuilder += data.slice(4).toString();

    this._packagesReceived++;

    if (this._packagesReceived === this._numberOfPackages) {
      this.message = this._stringBuilder;
    }
  }
}

module.exports = StringProtocol;